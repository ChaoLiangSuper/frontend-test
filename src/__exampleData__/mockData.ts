export const mockCoreData = {
  id: 'XXXXXX-ID',
  number: 'XXXXXX-NUM',
  createdDate: '2017-07-10 15:14:01',
  lastUpdateDate: '2019-05-27 19:56:17',
  type: 'type',
  source: 'source',
  state: 'state',
  shortDescription: 'short description',
  application: 'application',
  assignee: 'assignee',
};

export const mockServiceData = {
  test1: 'test1',
  test2: 'test2',
  test3: 'test3',
};

export const mockCardData = {
  coreData: mockCoreData,
  serviceData: mockServiceData,
};
