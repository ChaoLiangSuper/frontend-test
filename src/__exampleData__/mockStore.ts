import { createStore } from 'redux';

export const mockStore = createStore(() => ({ totalPage: 0, pages: {} }));
