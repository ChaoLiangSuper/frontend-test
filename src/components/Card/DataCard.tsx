import {
  Button,
  Card,
  CardActionArea,
  CardActions,
  CardContent,
  Typography,
} from '@material-ui/core';
import _ from 'lodash';
import React from 'react';
import { SidebarConsumer } from '../../contexts/SidebarContext';
import { ICardData } from '../../interfaces';
import useStyle from './style';

const DataCard = ({ cardData }: { cardData: ICardData }) => {
  const classes = useStyle();
  const { coreData } = cardData!;

  return (
    <SidebarConsumer>
      {(contextValue) => {
        const openSidebar = () => contextValue!.toggleSidebar(cardData!);
        return (
          <Card className={classes.card}>
            <CardActionArea onClick={openSidebar}>
              <CardContent>
                <Typography variant='body2' color='textSecondary' gutterBottom>
                  {_.capitalize(coreData.state)}
                </Typography>
                <Typography variant='h5'>{coreData.number}</Typography>
                <Typography variant='body2' color='textSecondary' noWrap>
                  {`Application: ${_.capitalize(coreData.application)}`}
                </Typography>
                <Typography
                  variant='body2'
                  color='textSecondary'
                  gutterBottom
                  noWrap
                >
                  {`Assignee: ${_.capitalize(coreData.assignee)}`}
                </Typography>
                <Typography variant='body2' className={classes.description}>
                  {coreData.shortDescription}
                </Typography>
              </CardContent>
            </CardActionArea>
            <CardActions>
              <Button onClick={openSidebar}>LEARN MORE</Button>
            </CardActions>
          </Card>
        );
      }}
    </SidebarConsumer>
  );
};

export default DataCard;
