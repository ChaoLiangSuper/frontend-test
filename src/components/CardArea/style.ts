import { makeStyles } from '@material-ui/core';

export default makeStyles({
  container: {
    display: 'flex',
    flexWrap: 'wrap',
    justifyContent: 'center',
    alignContent: 'flex-start',
    minHeight: '80vh',
    position: 'relative',
  },
});
