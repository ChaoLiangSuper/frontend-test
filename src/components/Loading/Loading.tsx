import { CircularProgress } from '@material-ui/core';
import React from 'react';
import useStyle from './style';

const Loading = () => {
  const classes = useStyle();

  return (
    <div className={classes.background}>
      <CircularProgress />
    </div>
  );
};

export default Loading;
