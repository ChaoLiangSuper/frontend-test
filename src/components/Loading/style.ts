import { makeStyles } from '@material-ui/core';

export default makeStyles(({ palette }) => ({
  background: {
    width: '100%',
    height: '100%',
    backgroundColor: palette.grey[100],
    position: 'absolute',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
  },
}));
