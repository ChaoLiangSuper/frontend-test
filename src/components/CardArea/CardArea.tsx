import _ from 'lodash';
import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { cardsPerPage } from '../../config';
import { ICardData, IStore } from '../../interfaces';
import { PAGE_FETCH } from '../../store/constants';
import Card, { EmptyCard } from '../Card';
import Loading from '../Loading';
import useStyle from './style';

export const CardArea = ({
  cards,
  getPageData,
  currentPage,
}: {
  cards?: ICardData[];
  getPageData: (n: number) => void;
  currentPage: number;
}) => {
  const classes = useStyle();

  useEffect(() => {
    getPageData(currentPage);
  }, [currentPage, getPageData]);

  return (
    <div className={classes.container}>
      {_.isNil(cards) ? (
        <Loading />
      ) : (
        <>
          {_.map(cards, (cardData, key) => (
            <Card cardData={cardData} key={key} />
          ))}
          {cards.length === cardsPerPage
            ? null
            : _.map(Array(cardsPerPage - cards.length), (value, key) => (
                <EmptyCard key={key} />
              ))}
        </>
      )}
    </div>
  );
};

export default connect(
  ({ pages }: IStore, { currentPage }: { currentPage: number }) => ({
    cards: pages[currentPage],
  }),
  (dispatch) => ({
    getPageData: _.debounce(
      (currentPage: number) => dispatch({ type: PAGE_FETCH, currentPage }),
      500
    ),
  })
)(CardArea);
