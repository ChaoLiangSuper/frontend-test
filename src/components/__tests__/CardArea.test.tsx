import { mount, shallow } from 'enzyme';
import _ from 'lodash';
import React from 'react';
import { Provider } from 'react-redux';
import { mockCardData } from '../../__exampleData__/mockData';
import { mockStore } from '../../__exampleData__/mockStore';
import DataCard, { EmptyCard } from '../Card';
import CardArea from '../CardArea';
import { CardArea as CardAreaComponent } from '../CardArea/CardArea';
import Loading from '../Loading';

describe('<CardArea /> component', () => {
  it('should successfully render without crash', () => {
    const wrapper = mount(
      <Provider store={mockStore}>
        <CardArea currentPage={1} />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render loading if cards are unavailable', () => {
    const wrapper = shallow(
      <CardAreaComponent getPageData={_.noop} currentPage={1} />
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Loading)).toHaveLength(1);
  });

  it('should render <Card /> and <EmptyCard />', () => {
    let wrapper = shallow(
      <CardAreaComponent
        cards={_.map(Array(10), () => mockCardData)}
        getPageData={_.noop}
        currentPage={1}
      />
    );

    expect(wrapper.find(DataCard)).toHaveLength(10);
    expect(wrapper.find(EmptyCard)).toHaveLength(2);

    wrapper = shallow(
      <CardAreaComponent
        cards={_.map(Array(12), () => mockCardData)}
        getPageData={_.noop}
        currentPage={1}
      />
    );

    expect(wrapper.find(EmptyCard)).toHaveLength(0);
  });
});
