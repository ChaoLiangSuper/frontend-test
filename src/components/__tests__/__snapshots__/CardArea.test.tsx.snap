// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardArea /> component should render loading if cards are unavailable 1`] = `
<div
  className="makeStyles-container-1"
>
  <Loading />
</div>
`;

exports[`<CardArea /> component should successfully render without crash 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(CardArea)
    currentPage={1}
  >
    <CardArea
      currentPage={1}
      getPageData={[Function]}
    >
      <div
        className="makeStyles-container-1"
      >
        <Loading>
          <div
            className="makeStyles-background-2"
          >
            <WithStyles(ForwardRef(CircularProgress))>
              <ForwardRef(CircularProgress)
                classes={
                  Object {
                    "circle": "MuiCircularProgress-circle",
                    "circleDisableShrink": "MuiCircularProgress-circleDisableShrink",
                    "circleIndeterminate": "MuiCircularProgress-circleIndeterminate",
                    "circleStatic": "MuiCircularProgress-circleStatic",
                    "colorPrimary": "MuiCircularProgress-colorPrimary",
                    "colorSecondary": "MuiCircularProgress-colorSecondary",
                    "indeterminate": "MuiCircularProgress-indeterminate",
                    "root": "MuiCircularProgress-root",
                    "static": "MuiCircularProgress-static",
                    "svg": "MuiCircularProgress-svg",
                  }
                }
              >
                <div
                  className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
                  role="progressbar"
                  style={
                    Object {
                      "height": 40,
                      "width": 40,
                    }
                  }
                >
                  <svg
                    className="MuiCircularProgress-svg"
                    viewBox="22 22 44 44"
                  >
                    <circle
                      className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                      cx={44}
                      cy={44}
                      fill="none"
                      r={20.2}
                      strokeWidth={3.6}
                      style={Object {}}
                    />
                  </svg>
                </div>
              </ForwardRef(CircularProgress)>
            </WithStyles(ForwardRef(CircularProgress))>
          </div>
        </Loading>
      </div>
    </CardArea>
  </Connect(CardArea)>
</Provider>
`;
